### PR TITLE
Feature/flatten permissions

### PR DESCRIPTION
**Pull request by jakobuid**
- Flatten all permissions to methods and events

### CODE REVIEW

File: walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/core/model/vo/sequence/SessionVO.kt [hunk 1]
```diff
@@ -20,13 +20,13 @@ internal data class SessionVO(
...
     val accounts: List<String> = emptyList(),
-    val chains: List<String>,
     val methods: List<String>,
-    val events: List<String>?,
+    val events: List<String>,
     val isAcknowledged: Boolean,
 ) : Sequence {
     val isPeerController: Boolean = peerParticipant?.keyAsHex == controllerKey?.keyAsHex
     val isSelfController: Boolean = selfParticipant.keyAsHex == controllerKey?.keyAsHex
+    val chains: List<String> get() = getChainIds(accounts)
```
Comment (TalhaAli00) [inline walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/core/model/vo/sequence/SessionVO.kt:29]:
```suggestion
    val chains: List<String>
        get() = accounts.map { accountId ->
            accountId.split(":").joinToString(":")
        }
```

File: walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/core/model/vo/sequence/SessionVO.kt [hunk 2]
```diff
@@ -73,12 +72,25 @@ internal data class SessionVO(
...
+                chains.add(getChainId(accountId))
+            }
+            return chains
+        }
+
+        private fun getChainId(accountId: String): String {
+            val elements = accountId.split(":")
+            val (namespace, reference) = Pair(elements[0], elements[1])
+            return "$namespace:$reference"
+        }
```
Comment (TalhaAli00) [inline walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/core/model/vo/sequence/SessionVO.kt:94]: with the above suggestion, you can remove these two functions.

File: walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/engine/model/mapper/EngineMapper.kt [hunk 3]
```diff
@@ -77,34 +72,26 @@ internal fun SessionParamsVO.EventParams.toEngineDOSessionEvent(topic: TopicVO):
...
 
 @JvmSynthetic
 internal fun SessionVO.toEngineDOApprovedSessionVO(): EngineDO.Session =
     EngineDO.Session(
-        topic, expiry,
-        accounts, selfMetaData?.toEngineDOAppMetaData(),
-        EngineDO.SessionPermissions(EngineDO.SessionPermissions.JsonRpc(methods), getEvents(events)),
-        EngineDO.Blockchain(chains),
+        topic, expiry, accounts, methods, events,
+        selfMetaData!!.toEngineDOAppMetaData()
```
Comment (TalhaAli00) [inline walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/engine/model/mapper/EngineMapper.kt:87]: using the `!!` can be dangerous. We should probably just throw an exception if metadata is null

File: walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/engine/model/mapper/EngineMapper.kt [hunk 4]
```diff
@@ -77,34 +72,26 @@ internal fun SessionParamsVO.EventParams.toEngineDOSessionEvent(topic: TopicVO):
...
 
 @JvmSynthetic
 internal fun SessionVO.toEngineDOSessionUpdateExpiry(expiryVO: ExpiryVO): EngineDO.SessionUpdateExpiry =
     EngineDO.SessionUpdateExpiry(
-        topic, expiryVO,
-        accounts, selfMetaData?.toEngineDOAppMetaData(),
-        EngineDO.SessionPermissions(EngineDO.SessionPermissions.JsonRpc(methods), getEvents(events)),
-        EngineDO.Blockchain(chains),
+        topic, expiryVO, accounts, methods, events,
+        selfMetaData!!.toEngineDOAppMetaData()
```
Comment (TalhaAli00) [inline walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/engine/model/mapper/EngineMapper.kt:94]: same as above

File: walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/engine/model/mapper/EngineMapper.kt [hunk 5]
```diff
@@ -119,43 +106,32 @@ internal fun PairingVO.toEngineDOSettledPairing(): EngineDO.PairingSettle =
 internal fun SessionVO.toSessionApproved(): EngineDO.SessionApproved =
     EngineDO.SessionApproved(
         topic.value,
-        peerMetaData?.toEngineDOMetaData(),
-        EngineDO.SessionPermissions(EngineDO.SessionPermissions.JsonRpc(methods),
-            if (events != null) EngineDO.SessionPermissions.Events(events) else null),
-        accounts
+        peerMetaData!!.toEngineDOMetaData(),
```
Comment (TalhaAli00) [inline walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/engine/model/mapper/EngineMapper.kt:109]: same as above

**Review by TalhaAli00 — APPROVED** [review]

File: walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/engine/model/mapper/EngineMapper.kt (hunk 3 above)
Comment (jakobuid) [reply walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/engine/model/mapper/EngineMapper.kt:87]: Ya true. However, the selfMetaData should not be nullable. I need to review this again.

File: walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/engine/model/mapper/EngineMapper.kt (hunk 4 above)
Comment (jakobuid) [reply walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/engine/model/mapper/EngineMapper.kt:94]: Same

File: walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/engine/model/mapper/EngineMapper.kt (hunk 5 above)
Comment (jakobuid) [reply walletconnectv2/src/main/kotlin/com/walletconnect/walletconnectv2/engine/model/mapper/EngineMapper.kt:109]: Same

**Pull request merged**